File: basics/TS/src/langchain/ChromaPdf.ts
import { BedrockEmbeddings } from "@langchain/community/embeddings/bedrock";
import { Bedrock } from "@langchain/community/llms/bedrock";
import { Chroma } from '@langchain/community/vectorstores/chroma'
import { PDFLoader } from 'langchain/document_loaders/fs/pdf'
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter'

const AWS_REGION = 'us-west-2'

const model = new Bedrock({
    model: 'amazon.titan-text-express-v1',
    region: AWS_REGION
})

const embeddings = new BedrockEmbeddings({
    region: AWS_REGION
})

const chromaConfig =         {
    collectionName: 'books',
    url: 'http://localhost:8000'
}

const question = "What themes does Gone with the Wind explore?";

async function main() {

    // const vectorStore = await loadAndStore();

    const vectorStore = await Chroma.fromExistingCollection(
        embeddings,
        chromaConfig
    )

    await queryStore(vectorStore);
}

async function loadAndStore(){
    // create the loader:
    const loader = new PDFLoader('assets/books.pdf', {
        splitPages: false
    });
    const docs = await loader.load();

    // split the docs:
    const splitter = new RecursiveCharacterTextSplitter({
        separators: [`. \n`]
    });

    const splittedDocs = await splitter.splitDocuments(docs);

    // store the data
    const vectorStore = await Chroma.fromDocuments(splittedDocs, embeddings, chromaConfig)
    await vectorStore.addDocuments(splittedDocs);

    return vectorStore;
}

async function queryStore(vectorStore: Chroma) {
    const retriever = vectorStore.asRetriever({
        k: 2
    });
    const results = await retriever.getRelevantDocuments(question);
    const resultDocs = results.map(
        result => result.pageContent
    )

    //build template:
    const template = ChatPromptTemplate.fromMessages([
        ['system', 'Answer the users question based on the following context: {context}'],
        ['user', '{input}']
    ]);

    const chain = template.pipe(model);

    const response = await chain.invoke({
        input: question,
        context: resultDocs
    });

    console.log(response)
}

main();

File: basics/PY/src/langchain/chroma_pdf.py
from langchain_community.llms import Bedrock
from langchain_community.embeddings import BedrockEmbeddings
from langchain_core.prompts import ChatPromptTemplate
from langchain_community.vectorstores import FAISS
from langchain_community.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import Chroma
import boto3

AWS_REGION = "us-west-2"

bedrock = boto3.client(service_name="bedrock-runtime", region_name=AWS_REGION)

model = Bedrock(model_id="amazon.titan-text-express-v1", client=bedrock)

bedrock_embeddings = BedrockEmbeddings(
    model_id="amazon.titan-embed-text-v1", client=bedrock
)

question = "What themes does Gone with the Wind explore?"


def load_and_store():
    loader = PyPDFLoader("assets/books.pdf")
    splitter = RecursiveCharacterTextSplitter(separators=["\n"])
    docs = loader.load()
    splitted_docs = splitter.split_documents(docs)
    vector_store = Chroma.from_documents(
        splitted_docs,
        embedding=bedrock_embeddings,
        collection_name="books",
    )
    vector_store.persist()
    return vector_store


def query_store(vector_store):
    retriever = vector_store.as_retriever(
        search_kwargs={"k": 2}  # maybe we can add a score threshold here?
    )
    results = retriever.get_relevant_documents(question)
    results_string = []
    for result in results:
        results_string.append(result.page_content)

    template = ChatPromptTemplate.from_messages(
        [
            (
                "system",
                "Answer the users question based on the following context: {context}",
            ),
            ("user", "{input}"),
        ]
    )
    chain = template.pipe(model)

    response = chain.invoke({"input": question, "context": results_string})
    print(response)


vector_store = load_and_store()
# vector_store = Chroma(
#     embedding_function=bedrock_embeddings,
#     collection_name="books",
# )
query_store(vector_store)
